ibis/udf/vectorized.py

```diff
@@ -38,7 +42,50 @@ def __call__(self, *args, **kwargs):
         # kwargs.
         @functools.wraps(self.func)
         def func(*args):
-            return self.func(*args, **kwargs)
+            # If cols are pd.Series, then we save and restore the index.
+            # Otherwise, we will reset the index to the natural index.
+            if hasattr(args[0], 'index'):
+                saved_index = args[0].index
+            else:
+                saved_index = None
+
+            result = self.func(*args, **kwargs)
+
+            if isinstance(self.output_type, dt.Struct):
+                if isinstance(result, pd.DataFrame) or (
```

Comment:
We should probably have function "coerce_struct_output" to capture this logic here

Reply:
Actually, why don't we do

```
if isinstance(self.output_type, dt.Struct):
    result = coerce_to_dataframe(result, self.output_type.names)
else:
    # coerce non struct output to standard form
```

Reply:
Sounds good, I can encapsulate some of this logic to make it cleaner.

Reply:
Before encapsulate, maybe explain first why checking array is needed here?

Reply:
Because if the intended return type of the UDF is an array, then we would mis-shape the result data here by coercing to DataFrame or Series.

For example, if `result` is
[np.array(1, 2, 3), np.array(4, 5, 6)] with column names 'A' and 'B'
and the output type of my UDF is Struct(Array, Array),
then what I want to return in my result is something like:
```
index       A           B
0       [1, 2, 3]   [4, 5, 6]
```
Without this check, we would instead return
```
index         A            B
0             1            4
1             2            5
2             3            6
```
The same is true for single-col UDFs.

Reply:
Can we pass the output type into coerce_to_dataframe to do the right thing?

Reply:
Unfortunately if we check the type directly inside this function, we get a circular dependency:

```
ibis/__init__.py:5: in <module>
    import ibis.expr.types as ir
ibis/expr/types.py:11: in <module>
    import ibis.util as util
ibis/util.py:27: in <module>
    import ibis.expr.datatypes as dt
ibis/expr/datatypes.py:146: in <module>
    class Null(DataType):
ibis/expr/datatypes.py:147: in Null
    scalar = ir.NullScalar
E   AttributeError: module 'ibis.expr.types' has no attribute 'NullScalar'
```

Reply:
Avoided the circular dependency by moving this function into schema.py